File: webapp.bodaqs.net/frontend/src/lib/api/preprocess.ts
export interface SignalsPayload {
  column_names: string[];
  n_rows: number;
  columns: Record<string, string>; // column name → base64 float32 LE
}

export interface PreprocessResponse {
  session_id: string;
  meta: Record<string, unknown>;
  source_sha256: string;
  signals: SignalsPayload;
  events: Record<string, unknown>[];
  metrics: Record<string, unknown>[];
  warnings: string[];
}

export interface PreprocessFormData {
  csv_file: File;
  bike_profile_json: File;
  sidecar_json: File;
  event_schema_yaml: File;
  preprocess_profile_json?: File;
}

export async function postPreprocess(data: PreprocessFormData): Promise<PreprocessResponse> {
  const form = new FormData();
  form.append("csv_file", data.csv_file);
  form.append("bike_profile_json", data.bike_profile_json);
  form.append("sidecar_json", data.sidecar_json);
  form.append("event_schema_yaml", data.event_schema_yaml);
  if (data.preprocess_profile_json) {
    form.append("preprocess_profile_json", data.preprocess_profile_json);
  }

  const response = await fetch("/api/preprocess", { method: "POST", body: form });
  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Preprocess failed (${response.status}): ${text}`);
  }
  return response.json() as Promise<PreprocessResponse>;
}

export function decodeSignalColumn(base64: string): Float32Array {
  const binary = atob(base64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return new Float32Array(bytes.buffer);
}

File: webapp.bodaqs.net/api/bodaqs_api/schemas/preprocess.py
"""Pydantic request/response models for the preprocess endpoint."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class SignalsPayload(BaseModel):
    """Processed signal columns encoded as base64 float32 arrays."""

    column_names: list[str] = Field(description="Ordered list of signal column names.")
    n_rows: int = Field(description="Number of samples in each column.")
    columns: dict[str, str] = Field(
        description="Map of column name → base64-encoded little-endian float32 array."
    )


class PreprocessResponse(BaseModel):
    """Response returned by POST /api/preprocess."""

    session_id: str = Field(description="Session identifier derived from the CSV filename stem.")
    meta: dict[str, Any] = Field(description="Session metadata (session['meta'] serialised to JSON).")
    source_sha256: str = Field(description="SHA-256 hex digest of the raw uploaded CSV bytes.")
    signals: SignalsPayload
    events: list[dict[str, Any]] = Field(
        description="Event detection results as records. NaN values are serialised as null."
    )
    metrics: list[dict[str, Any]] = Field(
        description="Metrics results as records. NaN values are serialised as null."
    )
    warnings: list[str] = Field(
        description="Pipeline warnings accumulated during preprocessing. Never empty-suppressed."
    )
